TDL-19892: Use common API calls for parent's data for parent/child streams

harshpatel4crest opened the pull request:
# Description of change
TDL-19892: Use common API calls for parent's data for parent/child streams
- Updated the code to use common API calls for parent-child streams.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

savan-chovatiya commented on tap_intercom/streams.py:
```diff
@@ -102,27 +141,66 @@ def sync(self,
+            is_child_selected = child_stream.tap_stream_id in self.selected_streams
+
+            if is_parent_selected and is_child_selected:
+                start_date = min(singer.utils.strptime_to_utc(parent_bookmark), singer.utils.strptime_to_utc(child_bookmark))
```
Change variable name here. This name is confusing in the later part with the `start_date` of config.

harshpatel4crest replied:
Updated the variable name.

savan-chovatiya commented on tap_intercom/sync.py:
```diff
@@ -49,13 +49,26 @@ def sync(config, state, catalog):
             stream_schema = stream.schema.to_dict()
             stream_metadata = metadata.to_map(stream.metadata)
 
+            if stream.tap_stream_id in list(CHILD_STREAMS.values()):
+                continue
```
Add a code comment here regarding the `continue` statement

harshpatel4crest replied:
Added comment.

savan-chovatiya commented on tests/unittests/test_conversation_part_bookmarks.py:
```diff
@@ -15,27 +31,32 @@ def test_conversation_parts_bookmarking(self, mocked_client_get, mocked_parent_r
         mocked_parent_records.return_value = [
-            {"id": 1, "updated_at": 1640636000}, # UTC datetime "2021-12-27T20:13:20.000000Z" for 1640636000
-            {"id": 2, "updated_at": 1640637000}  # UTC datetime "2021-12-27T20:30:00.000000Z" for 1640637000
+            {"id": 1, "updated_at": 1640636000000}, # UTC datetime "2021-12-27T20:13:20.000000Z" for 1640636000
+            {"id": 2, "updated_at": 1640637000000}  # UTC datetime "2021-12-27T20:30:00.000000Z" for 1640637000
```
The API call is the same as before so any reason to update the `updated_at` field's mock values?

harshpatel4crest replied:
The code will now go through `transform_times` when syncing conversation_parts, as a result, the parent's replication key will be in millisecond format.

dbshah1212 commented on tap_intercom/sync.py:
```diff
@@ -49,13 +49,26 @@ def sync(config, state, catalog):
+    # loop over the parent-child stream and add parent stream to sync if child stream is selected and parent stream is not selected
+    for parent_stream, child_stream in CHILD_STREAMS.items():
+        if child_stream in selected_stream_names and parent_stream not in selected_stream_names:
+            selected_streams.append(catalog.get_stream(parent_stream))
```
Adding parents into the selected stream is not the right thing to do, because we are not gonna write the parent's records. Create a separate list of sync_to_stream and inside that, you should add these parents.

harshpatel4crest replied:
Updated the code to add extra streams to sync in `streams_to_sync`.

savan-chovatiya commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,123 @@
+import singer
+
+# Mock class for stream schema
+class Schema:
```
The `Schema` mock class is already defined in `test_conversation_part_bookmarks.py` so please import it and use it.

harshpatel4crest replied:
Updated the test case.

savan-chovatiya commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,123 @@
+        self.metadata = {}
+
+# Mock class for catalog
+class Catalog:
```
The `Stream` and `Catalog` mock classes are already defined in `test_conversation_part_bookmarks.py` so please import it and use it.

harshpatel4crest replied:
Used the already defined mock Catalog class.

dbshah1212 commented on tap_intercom/sync.py:
```diff
@@ -49,13 +49,26 @@ def sync(config, state, catalog):
+    for stream in selected_streams:
+        selected_stream_names.append(stream.stream)
+
+    # loop over the parent-child stream and add parent stream to sync if child stream is selected and parent stream is not selected
```
```suggestion
    # Loop over the parent-child stream and add the parent stream to sync if the child stream is selected and the parent stream is not selected
```

harshpatel4crest replied:
Updated the comment.

dbshah1212 commented on tap_intercom/sync.py:
```diff
@@ -49,13 +49,26 @@ def sync(config, state, catalog):
+    selected_stream_names = []
+    selected_streams = list(catalog.get_selected_streams(state))
+    for stream in selected_streams:
+        selected_stream_names.append(stream.stream)
```
Why we are not adding `stream.tap_stream_id`?

harshpatel4crest replied:
Most of the time tap_stream_id and stream will have the same values. Updated the current code to use tap_stream_id.

savan-chovatiya commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,123 @@
+
+    @parameterized.expand([
+        ['parent_child_selected_start_date', [['conversations', 'conversation_parts'], {}], '2021-01-01'],
+        ['parent_child_selected_state', [['conversations', 'conversation_parts'], {'bookmarks': {'conversation_parts': {'updated_at': '2021-01-02'}, 'conversations': {'updated_at': '2021-01-03'}}}], '2021-01-02'],
+        ['parent_selected_start_date', [['conversations'], {}], '2021-01-01'],
+        ['parent_selected_state', [['conversations'], {'bookmarks': {'conversation_parts': {'updated_at': '2021-01-02'}, 'conversations': {'updated_at': '2021-01-03'}}}], '2021-01-03'],
+        ['child_selected_start_date', [['conversation_parts'], {}], '2021-01-01'],
+        ['child_selected_state', [['conversation_parts'], {'bookmarks': {'conversation_parts': {'updated_at': '2021-01-02'}, 'conversations': {'updated_at': '2021-01-03'}}}], '2021-01-02'],
```
Divide long lines into multiline for better readability.

harshpatel4crest replied:
Updated.

savan-chovatiya commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,123 @@
+        ['child_selected', ['conversation_parts'], {'stream': 'conversations'}]
+    ])
+    @mock.patch('tap_intercom.streams.IncrementalStream.sync', return_value = {})
+    def test_sync(self, name, test_data, expected_data, mocked_sync):
```
Give a more explanatory test case name.

harshpatel4crest replied:
Updated the test case name.

savan-chovatiya commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,123 @@
+    ])
+    @mock.patch('singer.write_schema')
+    @mock.patch('tap_intercom.streams.Conversations.get_records')
+    def test_syncing(self, name, test_data, expected_data, mocked_get_records, mocked_write_schema):
```
Give a more explanatory test case name.

harshpatel4crest replied:
Updated the test case name.

dbshah1212 commented on tap_intercom/sync.py:
```diff
@@ -4,7 +4,7 @@
 
 from tap_intercom.client import IntercomClient
-from tap_intercom.streams import STREAMS
+from tap_intercom.streams import STREAMS, CHILD_STREAMS
```
Why do we require CHILD_STREAMS dict, we can define them in stream class?

harshpatel4crest replied:
Yes, we can define it in the stream's class, but it will become easy to loop on CHILD_STREAMS rather than STREAMS for preparing the `streams_to_sync` list in the sync.py file.

dbshah1212 replied:
We shoud include the selected streams in streams_to_sync so run a loop on streams and check if stream is selected or child stream of it is selected then add that stream in streams_to_sync.

harshpatel4crest replied:
We are looping in the selected streams to use the shuffling functionality of singer-python for interrupted syncs and adding the independent streams and parent stream in **streams_to_sync**

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -131,6 +209,11 @@ def sync(self,
                     counter.increment()
                     max_datetime = max(record_datetime, max_datetime)
+
+                # sync child stream, if child is selected
```
```suggestion
                # Sync child stream, if the child is selected
```

harshpatel4crest replied:
Updated the comment.

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -102,27 +141,66 @@ def sync(self,
                     )
 
-                if record_datetime >= bookmark_datetime:
+                # write record if parent is selected
```
```suggestion
                # Write record if a parent is selected
```

harshpatel4crest replied:
Updated comment.

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -102,27 +141,66 @@ def sync(self,
+            child_schema = child_stream_.schema.to_dict()
+            child_metadata = metadata.to_map(child_stream_.metadata)
+            if is_child_selected:
+                # write schema for child stream as it will be synced by the parent stream
```
```suggestion
                # Write schema for child stream as it will be synced by the parent stream
```

dbshah1212 replied:
Make first latter capital.

harshpatel4crest replied:
Updated comment.

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -102,27 +141,66 @@ def sync(self,
+        is_parent_selected = True
+        is_child_selected = False
+
+        # if the current stream has child stream, then get the child stream's bookmark and
```
```suggestion
        # If the current stream has a child stream, then get the child stream's bookmark and
```

harshpatel4crest replied:
Updated comments.

harshpatel4crest commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,123 @@
+@mock.patch('tap_intercom.streams.Conversations.get_records', side_effect = [[{'id': 1, 'updated_at': 1609804800000}]]) # '2021-01-05'
+@mock.patch('tap_intercom.streams.transform', side_effect = transform)
+@mock.patch('tap_intercom.streams.BaseStream.sync_substream')
+class TestParentChildWriteRecords(unittest.TestCase):
```
Added unittest for the condition.

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -102,35 +141,79 @@ def sync(self,
-        LOGGER.info("Stream: {}, initial max_bookmark_value: {}".format(self.tap_stream_id, start_date))
-        bookmark_datetime = singer.utils.strptime_to_utc(start_date)
-        max_datetime = bookmark_datetime
+        # check if the current stream has child stream or not
```
Make the first latter of the sentence Capital for each comment.

harshpatel4crest replied:
Updated comments.

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -102,35 +141,79 @@ def sync(self,
-        bookmark_datetime = singer.utils.strptime_to_utc(start_date)
-        max_datetime = bookmark_datetime
+        # check if the current stream has child stream or not
+        has_child = self.tap_stream_id in CHILD_STREAMS
```
Remove CHILD_STREAMS dict, we can handle it using parent and child classes.

harshpatel4crest replied:
Updated the code.

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -684,3 +697,8 @@ def get_records(self, bookmark_datetime=None, is_parent=False) -> Iterator[list]
     "teams": Teams
 }
+
+CHILD_STREAMS = {
```
handle it using classes.

harshpatel4crest replied:
Updated.

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -60,7 +63,7 @@ def get_parent_data(self, bookmark_datetime: datetime = None) -> list:
         :return: A list of records
         """
```
get_parent_data method called only once and it doesn't make sense as part of the parent-child we are restricting parent call to once, so can we remove this or move this to the appropriate class?

dbshah1212 commented on tap_intercom/streams.py:
```diff
@@ -224,6 +296,18 @@ class Admins(FullTableStream):
     path = 'admins/{}'
     parent = AdminList
```
Why child var is not defined for `AdminList`? We should call Admins in `AdminList` calls, not the reverse one.

dbshah1212 replied:
If it is scenario at other place then also update it.

harshpatel4crest replied:
The AdminList and Admin are parent-child streams, but the **admin** is written in the catalog but the admin_list is not. The purpose of admin_list is to return a list of admins through which individual admins will be synced.

dbshah1212 replied:
Can you mention the same in the comment?

harshpatel4crest replied:
Added comment in sync.py file.

RushiT0122 commented on tap_intercom/streams.py:
```diff
@@ -102,35 +130,79 @@ def sync(self,
+                sync_start_date = singer.utils.strptime_to_utc(parent_bookmark)
+            elif is_child_selected:
+                sync_start_date = singer.utils.strptime_to_utc(child_bookmark)
+            else:
+                return state
```
Is this else block redundant since sync is called when parent or child is selected so either of above 3 conditions will be true?

harshpatel4crest replied:
Yes, the Tap will not go into the else condition.

RushiT0122 commented on tap_intercom/sync.py:
```diff
@@ -40,19 +40,46 @@ def translate_state(state):
+            streams_to_sync.append(stream)
+        # If the stream is child stream, then add parent stream to sync
+        elif stream_obj.parent:
+            # The AdminList and Admin are parent-child streams, the admin is written in the catalog but the admin_list is not.
```
Correction: # The AdminList and Admin are parent-child streams, the `admins` is written in the catalog but the admin_list is not.

harshpatel4crest replied:
Updated the comment.

RushiT0122 commented on tests/unittests/test_conversation_part_bookmarks.py:
```diff
@@ -15,27 +41,32 @@ def test_conversation_parts_bookmarking(self, mocked_client_get, mocked_parent_r
 
         # Expected call of write_bookmark() function
-        state = {'bookmarks': {'conversation_parts': {'updated_at': '2021-12-27T20:30:00.000000Z'}}}
+        state = {'bookmarks': {'conversation_parts': {'updated_at': '2021-12-27T20:30:00.000000Z'}, 'conversations': {'updated_at': '2021-12-27T20:30:00.000000Z'}}}
```
```suggestion
        state = {'bookmarks': {'conversation_parts': {'updated_at': '2021-12-27T20:30:00.000000Z'},
                               'conversations': {'updated_at': '2021-12-27T20:30:00.000000Z'}}}
```

RushiT0122 commented on tests/unittests/test_conversation_part_bookmarks.py:
```diff
@@ -15,27 +41,32 @@ def test_conversation_parts_bookmarking(self, mocked_client_get, mocked_parent_r
+        # Initialize IntercomClient and Conversations object
         client = IntercomClient('dummy_token', None)
-        conversation_part = ConversationParts(client)
+        conversation_part = Conversations(client, Catalog(['conversations', 'conversation_parts']), ['conversations', 'conversation_parts'])
```
```suggestion
        conversation_part = Conversations(client=client, 
                                            catalog=Catalog(['conversations', 'conversation_parts']), 
                                            selected_streams=['conversations', 'conversation_parts'])
```

harshpatel4crest replied:
Updated.

RushiT0122 replied:
I can't see the changes been made.

RushiT0122 commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,110 @@
+            Test case to verify we set expected start date to sync for parent-child streams
+        """
+        client = IntercomClient('test_access_token', 300)
+        conversations = Conversations(client, Catalog(test_data[0]), test_data[0])
```
```suggestion
        conversations = Conversations(client=client, catalog=Catalog(test_data[0]), selected_streams=test_data[0])
```

harshpatel4crest replied:
Updated.

RushiT0122 replied:
I can't see the changes been made.

RushiT0122 commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,110 @@
+        conversations = Conversations(client, Catalog(test_data[0]), test_data[0])
+        state = test_data[1]
+        config = {'start_date': '2021-01-01'}
+        conversations.sync(state, {}, {}, config, None)
```
```suggestion
        conversations.sync(state=state, stream_schema={}, stream_metadata={}, config=config, transformer=None)
```

This will give better readability.

harshpatel4crest replied:
Updated the code.

RushiT0122 replied:
I can't see the changes been made.

RushiT0122 commented on tests/unittests/test_parent_child_sync.py:
```diff
@@ -0,0 +1,110 @@
+            Test case to verify we write parent records if the stream is selected
+        """
+        client = IntercomClient('test_access_token', 300)
+        conversations = Conversations(client, Catalog(['conversations']), ['conversations'])
+        config = {'start_date': '2021-01-01'}
+        conversations.sync({}, {}, {}, config, None)
```
```suggestion
        conversations = Conversations(client=client, catalog=Catalog(['conversations']), selected_streams=['conversations'])
        config = {'start_date': '2021-01-01'}
        conversations.sync(state={}, stream_schema={}, stream_metadata={}, config=config, transformer=None)
```
For better readability.

harshpatel4crest replied:
Updated as per suggestion.

harshpatel4crest commented:
@RushiT0122 I have already made the suggested changes in unittests for argument passing.
Reference: https://github.com/singer-io/tap-intercom/pull/55/files#diff-6fe7cea6821dd5f9f6c736a26c7a59510fae9cc9e515c5818343c63c587d5dc0R57 and https://github.com/singer-io/tap-intercom/pull/55/files#diff-98d50c077e9f69b95074f17c78efecc96760893b122bde89e3da1758e8ef025bR52

harshpatel4crest commented on tap_intercom/sync.py:
```diff
@@ -40,19 +40,46 @@ def translate_state(state):
 
+def get_streams_to_sync(catalog, selected_streams, selected_stream_names):
+    """
+        Get streams to sync ie. selected streams and parent stream if child stream stream or parent stream is selected
```
Updated comment.

harshpatel4crest commented:
@RushiT0122 Removed redundant code for returning state when parent and child streams are not selected.

kspeer825 reviewed (changes requested):
The current implementation of the bookmarks test for tap-intercom is invalid. It combines the standard tap-tester start date and bookmarks tests. It is not actually providing coverage and needs to be addressed for this change .

Also this change should have a tap-tester test that performs syncs for the following scenario:
 - child bookmark is earlier than parent bookmark
 - parent bookmark is earlier than child bookmark
 and should assert that:
 - in both cases the earlier bookmark is used by looking at the replicated records.
 - in both cases records between the bookmark values are replicated.

harshpatel4crest commented:
> The current implementation of the bookmarks test for tap-intercom is invalid. It combines the standard tap-tester start date and bookmarks tests. It is not actually providing coverage and needs to be addressed for this change .
> 
> Also this change should have a tap-tester test that performs syncs for the following scenario:
> 
> * child bookmark is earlier than parent bookmark
> * parent bookmark is earlier than child bookmark
>   and should assert that:
> * in both cases the earlier bookmark is used by looking at the replicated records.
> * in both cases records between the bookmark values are replicated.

Thanks, @kspeer825 for the suggestion. We updated the bookmark test. As we are writing the parent's replication key in the child stream's bookmark, so we have skipped the conversation_parts(child stream) but we have included conversations(parent) in the current bookmark test case by setting the dynamic start date as the stream was being skipped earlier due to more data.

We are collecting data from minimum bookmarks out of parent-child streams, but tap will only write data that is greater than the bookmark for the parent stream and all the child for the parent greater than the child's bookmark. The final output of the tap will remain the same after this change.
We added a new integration test for the scenario mentioned above. We will also make sure to add this test in all the taps from now onwards.
